### ISSUE TITLE
[kmac] SCA hardening follow-up tasks

### ISSUE THREAD

**Issue by vogelpi** [lowRISC/opentitan#12323]
During the review of #12194 , we've identified a couple of follow-up tasks that should be taken care of:
- [ ] Enhance prim_edn_req to omit the packing FIFO if the input and output entropy widths match. See https://github.com/lowRISC/opentitan/pull/12194#discussion_r853241723
- [ ] Add hook to trigger a complete reseeding of the internal PRNG on demand. Right now to perform a complete reseed, software has to trigger 5 subsequent reseeding operations via the CMD register. See https://github.com/lowRISC/opentitan/pull/12194#discussion_r853251786
- [x] Re-investigate potential leakage upon loading the key without software-based key remasking. We are not sure if this was just an artifact due to the measurement method. See https://github.com/lowRISC/opentitan/pull/12194#discussion_r855580533
  - FYI, @vrozic ran an experiment to verify software-based key remasking is indeed required. It's now enabled for SCA experiments here [#13298](https://github.com/lowRISC/opentitan/pull/13298)
- [ ] Not strictly KMAC related: Write up a separate SCA hardening doc summarizing the rationale behind the implemented changes and to give guidance to others.

**Comment by cdgori** [comment]
In #12194 I was asking about where the seeds came from, you found a cut-paste/same-seed error, but I think the question still stands, so maybe this is another checkbox item for this issue: 

Perhaps I just don't know the flow and this already happens, but it would be good if there was a way to (consistently) re-derive and/or update the seeds used here, ideally from some base/root seed.

**Comment by vogelpi** [comment]
If the user doesn't provide a custom seed to the `./util/design/gen-lfsr-seed.py` script, it will reseed the [internal Mersenne Twister PRNG of Python with the current system time](https://docs.python.org/3/library/random.html). I think this is fine as long as all these constants are indeed exposed as parameters at the top level where we use the rnd_cnst flow implemented by @msfschaffner to re-generate the parameters for a specific top level.  The goal of the `SEC_RND_CNST` checklist item is to ensure that indeed this flow is used for all compile-time random netlist constants of an IP.

When regenerating these random netlist constants for the top level, topgen does the following, it either uses a seed value provided as command line argument, takes the value from the top hjson, or alternatively again the system time. Internally it also uses the Mersenne Twister PRNG of Python. @cdgori , do you think we should switch to something stronger, cryptograhpically more secure? For the top-level flow?

**Comment by vogelpi** [comment]
The first point (prim_edn_req) is not critical. It would allow to align the RTL with other entropy consumers somehwat. But this can be deferred.

The second point (reseed entire PRNG on demand) is not critical as well IMO. If software wants to perform a complete reseed meaning all cascaded LFSRs, it can do so by triggering five subsequent reseeds. So I think it's also not urgent unless @eunchan, @tjaychen or you think differently @moidx . 

The third point has been addressed and resolved now.

The fourth point isn't critical as it's documentation only.

**Comment by vogelpi** [comment]
Everything that is critical from this issue has been done already and the remaining sub tasks will most likely be moved to backlog. I am also unassigning Vladimir and myself.

**Comment by ballifatih** [comment]
Triaged for `kmac`. I think these changes are not critical for M2.5. Let me know if you think otherwise @vogelpi.

**Comment by vogelpi** [comment]
> During the review of #12194 , we've identified a couple of follow-up tasks that should be taken care of:
> 
>     * [ ]  Enhance prim_edn_req to omit the packing FIFO if the input and output entropy widths match. See [[kmac] Rework Keccak/SHA3 core and PRNG for SCA hardening #12194 (comment)](https://github.com/lowRISC/opentitan/pull/12194#discussion_r853241723)
> 
The following IPs in Earlgrey currently employ prim_edn_req primitives
- alert_handler: 1 prim_edn_req with matching input and output width of 32 bits
- otbn: 2 prim_edn_req with non-matching input and output width (32 bits in, 256 bits out)
- keymgr: 1 prim_edn_req of non-matching input and output width (32 bits in, 64 bits out)
- otp_ctrl: 1 prim_edn_req of non-matching input and output width (32 bits in, 64 bits out)
- rv_core_ibex:  1 prim_edn_req with matching input and output width of 32 bits

So by doing this change we could save 3 x 32 FFs = 96 FFs for Earlgrey. The effort for doing this seems not justified to me given we have other, more relevant things to do for PROD and the associated risk of messing something up. Remember: these are CDCs, it might be good to have a set of FFs before consuming the received entropy downstream.  

>     * [ ]  Add hook to trigger a complete reseeding of the internal PRNG on demand. Right now to perform a complete reseed, software has to trigger 5 subsequent reseeding operations via the CMD register. See [[kmac] Rework Keccak/SHA3 core and PRNG for SCA hardening #12194 (comment)](https://github.com/lowRISC/opentitan/pull/12194#discussion_r853251786)
> 

I will need to rework the KMAC reseeding slightly as part of reworking the masking PRNG anyway. Most likely we will have less state and reseeding always everything might be less costly but is safer. We don't need to track this separately 

>     * [x]  Re-investigate potential leakage upon loading the key without software-based key remasking. We are not sure if this was just an artifact due to the measurement method. See [[kmac] Rework Keccak/SHA3 core and PRNG for SCA hardening #12194 (comment)](https://github.com/lowRISC/opentitan/pull/12194#discussion_r855580533)
>
>       * FYI, @vrozic ran an experiment to verify software-based key remasking is indeed required. It's now enabled for SCA experiments here [#13298](https://github.com/lowRISC/opentitan/pull/13298)
> 
Done.

>     * [ ]  Not strictly KMAC related: Write up a separate SCA hardening doc summarizing the rationale behind the implemented changes and to give guidance to others.

I'll file a seperate documentation issue on this topic.

To summarize: it's fine to close this issue.

**Comment by johannheyszl** [comment]
removing Hotlist label bc closed
